### PR TITLE
spirv-fuzz: Get order right for OpSelect arguments

### LINKED ISSUE #3948: spirv-fuzz: FlattenConditionalBranch does not handle OpPhi to Select properly when condition can branch straight to convergence block
When the conditional being flattened can branch straight to the convergence block, the ids consumed by an OpPhi can appear in the wrong order when the OpPhi is turned into an OpSelect.

### PR DESCRIPTION
**Pull request by afd**
Fixes #3948.

### CODE REVIEW

File: source/fuzz/transformation_flatten_conditional_branch.cpp [hunk 1]
```diff
@@ -843,85 +843,123 @@ void TransformationFlattenConditionalBranch::
...
+    // predecessor is associated with the "true" path of the conditional
+    // we are flattening, the other with the "false" path, but these
+    // predecessors can appear in either order as operands to the OpPhi
+    // instruction.  We determine in which order the OpPhi inputs should
+    // appear as OpSelect arguments by checking dominance of the true and
+    // false immediate successors of the header block.
+    if (branch_instruction_true_block_id == convergence_block_id) {
+      // The branch instruction's true block is the convergence block.  This
+      // means that the OpPhi's value associated with the branch instruction's
+      // block should the the "true" result of the OpSelect.
```
Comment (paulthomson) [inline source/fuzz/transformation_flatten_conditional_branch.cpp:917]: Typo: the the

File: source/fuzz/transformation_flatten_conditional_branch.cpp [hunk 2]
```diff
@@ -843,85 +843,123 @@ void TransformationFlattenConditionalBranch::
...
+      assert(branch_instruction_false_block_id != convergence_block_id &&
+             "Control should not reach here if both branches target the "
+             "convergence block.");
+      if (phi_inst->GetSingleWordInOperand(1) == message_.header_block_id()) {
+        operands.emplace_back(phi_inst->GetInOperand(0));
+        operands.emplace_back(phi_inst->GetInOperand(2));
+      } else {
+        assert(phi_inst->GetSingleWordInOperand(3) ==
+                   message_.header_block_id() &&
+               "The convergence block has the header block as a predecessor.");
```
Comment (paulthomson) [inline source/fuzz/transformation_flatten_conditional_branch.cpp:927]: Is this message correct? I don't 100% understand it.

File: source/fuzz/transformation_flatten_conditional_branch.cpp [hunk 3]
```diff
@@ -843,85 +843,123 @@ void TransformationFlattenConditionalBranch::
...
+        operands.emplace_back(phi_inst->GetInOperand(2));
+        operands.emplace_back(phi_inst->GetInOperand(0));
+      }
+    } else if (branch_instruction_false_block_id == convergence_block_id) {
+      // The branch instruction's false block is the convergence block.  This
+      // means that the OpPhi's value associated with the branch instruction's
+      // block should the the "false" result of the OpSelect.
```
Comment (paulthomson) [inline source/fuzz/transformation_flatten_conditional_branch.cpp:934]: Typo again: the the

File: source/fuzz/transformation_flatten_conditional_branch.cpp [hunk 4]
```diff
@@ -843,85 +843,123 @@ void TransformationFlattenConditionalBranch::
...
+      if (phi_inst->GetSingleWordInOperand(1) == message_.header_block_id()) {
+        operands.emplace_back(phi_inst->GetInOperand(2));
+        operands.emplace_back(phi_inst->GetInOperand(0));
+      } else {
+        assert(phi_inst->GetSingleWordInOperand(3) ==
+                   message_.header_block_id() &&
+               "The convergence block has the header block as a predecessor.");
```
Comment (paulthomson) [inline source/fuzz/transformation_flatten_conditional_branch.cpp:941]: Again: is this message correct?

File: source/fuzz/transformation_flatten_conditional_branch.cpp [hunk 5]
```diff
@@ -843,85 +843,123 @@ void TransformationFlattenConditionalBranch::
...
+    uint32_t branch_instruction_false_block_id =
+        branch_instruction.GetSingleWordInOperand(2);
+
+    // The OpPhi takes values from two distinct predecessors.  One
+    // predecessor is associated with the "true" path of the conditional
+    // we are flattening, the other with the "false" path, but these
+    // predecessors can appear in either order as operands to the OpPhi
+    // instruction.  We determine in which order the OpPhi inputs should
+    // appear as OpSelect arguments by checking dominance of the true and
+    // false immediate successors of the header block.
```
Comment (paulthomson) [inline source/fuzz/transformation_flatten_conditional_branch.cpp:913]: Maybe update this comment slightly to include that we first check if either immediate successor is the convergence block.

**Review by paulthomson — APPROVED** [review]

File: source/fuzz/transformation_flatten_conditional_branch.cpp (hunk 1 above)
Comment (afd) [reply source/fuzz/transformation_flatten_conditional_branch.cpp:917]: Fixed.

File: source/fuzz/transformation_flatten_conditional_branch.cpp (hunk 3 above)
Comment (afd) [reply source/fuzz/transformation_flatten_conditional_branch.cpp:934]: Fixed.

File: source/fuzz/transformation_flatten_conditional_branch.cpp (hunk 2 above)
Comment (afd) [reply source/fuzz/transformation_flatten_conditional_branch.cpp:927]: I tried to phrase it better.

File: source/fuzz/transformation_flatten_conditional_branch.cpp (hunk 4 above)
Comment (afd) [reply source/fuzz/transformation_flatten_conditional_branch.cpp:941]: Also re-phrased.

File: source/fuzz/transformation_flatten_conditional_branch.cpp (hunk 5 above)
Comment (afd) [reply source/fuzz/transformation_flatten_conditional_branch.cpp:913]: Done.

**Pull request merged**